### DISCUSSION TITLE
How can I change nvim-treesitter to use my own fork?

### DISCUSSION THREAD

**Discussion by markus-oberhumer-forks** [LazyVim/LazyVim#452] [Q&A] [answered]
I'd like to override the default url and branch of nvim-treesitter/nvim-treesitter.

As this must be done quite early I'm unsure where to put that code - I tried a number of things including something like

```lua
local ts = require("lazy.core.config").spec.plugins["nvim-treesitter"]
ts.url = "https://github.com/markus-oberhumer-forks/nvim-treesitter--nvim-treesitter"
ts.branch = "mfx"
```

in lua/config/ts.lua but nothing worked.

Thanks for any help!

**Comment by folke** [comment]
You need to clone you fork yourself. Check the config.dev options

**Comment by markus-oberhumer-forks** [comment]
I had tried something like that, but I also have to override `config.name` to exactly match the original plugin, and then lazy.nvim complains:

`Two plugins with the same name and different url:`

**Reply by folke** [reply]
You just keep the regular treesitter name in your specs. That's what the dev option does. Just read the docs please. It's all in there.

```lua
  {
    "nvim-treesitter/nvim-treesitter",
   dev = true,
   -- ... your other options
}
```

Make sure to configure `config.dev.path` in lazy.nvim

**Comment by markus-oberhumer-forks** [comment]
I'm starting with a completely fresh LazyVim/starter Git repo, just adding one file

lua/plugins/treesitter.lua

```lua
return {
  {
    "nvim-treesitter/nvim-treesitter",
    dev = true,
    name = "nvim-treesitter",
    url = "https://github.com/markus-oberhumer-forks/nvim-treesitter--nvim-treesitter",
    branch = "mfx",
  },
}
```

Now lazy.nvim complains about "Two plugins with the same name and different url".

And if I remove the `"name ="` line then my url is completely ignored.

**Reply by folke** [reply]
Don't add `url=`! Again, `dev=true` takes care of that! and also don't add `branch`

Just clone your fork manually in your projects folder and **ONLY** set `dev=true`

**Reply by markus-oberhumer-forks** [reply]
Ok, I see - I'd guess what I'd really need would be some `force_url` option in lazy.nvim.

Anyway, **many thanks** for your help and all your great work!

**Reply by folke** [reply]
why would you need a force_url option?? You simply can't have multiple specs that have a different url for the same name.

Developing plugins should be done with the `dev=true` option, or using the `config.dev.patterns`

**Reply by markus-oberhumer-forks** [reply]
Well, I'm not developing a plugin, I just want to use my fork (which has updated tree-sitter injections and queries) and had hoped that I can easily override the plugin url & branch.

**Reply by amaanq** [reply]
I'm so confused - your fork is just effectively a mirror so why not use the main repo to begin with...?

**Reply by markus-oberhumer-forks** [reply]
My fork contains the `force_url` enhancement which I'm currently (successfully) testing on my local machines:

See  https://github.com/markus-oberhumer-forks/folke--lazy.nvim/commit/56248c74a3948d9d2a6cdf92b4c269405f2bfb65

I'm not sure if I will be able to convince @folke to merge this, but maybe I just poorly explained my use case.


## Test usage: 

apply the following diff to https://github.com/LazyVim/starter


```diff
diff --git a/lua/config/lazy.lua b/lua/config/lazy.lua
index 891b190..5ec11dd 100644
--- a/lua/config/lazy.lua
+++ b/lua/config/lazy.lua
@@ -1,13 +1,28 @@
 local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
 if not vim.loop.fs_stat(lazypath) then
   -- bootstrap lazy.nvim
+  -- vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", "--branch=stable", lazypath })
+  -- using fork with force_url support
   -- stylua: ignore
-  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", "--branch=stable", lazypath })
+  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/markus-oberhumer-forks/folke--lazy.nvim", "--branch=force_url", lazypath })
 end
 vim.opt.rtp:prepend(vim.env.LAZY or lazypath)

 require("lazy").setup({
   spec = {
+    -- using forks with force_url
+    {
+      "folke/lazy.nvim",
+      name = "lazy.nvim",
+      force_url = "https://github.com/markus-oberhumer-forks/folke--lazy.nvim",
+      force_branch = "force_url",
+    },
+    {
+      "nvim-treesitter/nvim-treesitter",
+      name = "nvim-treesitter",
+      force_url = "https://github.com/markus-oberhumer-forks/nvim-treesitter--nvim-treesitter",
+      force_branch = "mfx",
+    },
     -- add LazyVim and import its plugins
     { "LazyVim/LazyVim", import = "lazyvim.plugins" },
     -- import any extras modules here
```

**Reply by folke** [reply]
Nope, not planning to merge this. As explained, you can use the `dev` option for this. It doesn't matter whether you develop the plugin or not. But you do you. Create a fork of LazyVim, to install a fork of lazy to manage a fork of nvim-treesitter. Sounds great.

**Reply by markus-oberhumer-forks** [reply]
I cannot use `dev` because that does disable the automatic update/management of the affected plugin, and I need the exactly same setup being automated on dozens of (virtual) machines.

**Reply by rameshsanth** [reply]
Folke, please consider allowing NVIM plugin development without having to tinker with Lazy.
A developer should be able to focus on the plugin they're developing and shouldn't have to start mucking around with Lazy plugin to get their plugin.

**Reply by Jomik** [reply]
@rameshsanth Well, you can just decide to not have lazy manage your plugin then? 🤔 
Lazy does make it quite easy to develop your plugin, as you can simply toggle the `dev` field..
I understand why Folke does not want to merge this, as it can quickly become quite hard to understand where a plugin comes from, as specs can start including a different url from what you would expect from the name. Loading order of the plugins would suddenly start to matter quite a bit, which is something I think we should avoid as much as possible..

The only real option in my head here would be to have an entirely different location in your Lazy config, where you can map plugin names to git urls, and that would be a complete override. Then it would not be part of the plugin spec.
You could open a discussion if you think there is a super great use-case https://github.com/folke/lazy.nvim/discussions

But I would argue that you should probably just copy the plugin spec and then point that at your fork - what happens if LazyVim updates the spec with something your fork is not up-to-date with? Your config would break 😄

**Reply by rameshsanth** [reply]
Jomik,
As a long time user of Packer who recently switched to lazy and now trying out LazyVim, I'm well aware of the other plugin managers. The recommendation to not use LazyVim is unhelpful and exactly the sort of attitude that kills innovation and drives away possible contributors.

I genuinely believe the plugin manager should not get in the way of Nvim plugin development, such as having to fork plugin manager's repo and make changes in order to develop other plugin.

I admit I'm not familiar enough of using/configuring LazyVim, let alone understanding its design and inner working. 
Therefore I cannot brainstorm or suggest how this can be implemented easier. 

However I think I'm starting to realize that LazyVim is not a plugin manager (Lazy is). So, perhaps requests such as these are not high priority 

Thanks for reading and responding.

**Reply by amaanq** [reply]
Except you can just set `dev = true` in the plugin's spec..I just don't see why you wouldn't have it cloned locally if you're developing a plugin

**Reply by Jomik** [reply]
@rameshsanth, as @amaanq says, it is because it is super simple to develop a plugin with `lazy.nvim`. You simply have to [define a folder where your development plugins live](https://github.com/Jomik/dotfiles/blob/bcbe995fdcf0a7af3bc62cc2e97c40dbec13be27/.config/nvim/lua/config/lazy.lua#L37), and then set `dev = true` on the `PluginSpec`.
I think @markus-oberhumer-forks's problem lies in the fact that they are not simply developing a plugin, they have a fork that they want to use in "production".
The only solution there, is to then explicitly use the fork, not try and mangle one plugin spec to match the other - as I mentioned, they could easily drift apart.

**Reply by markus-oberhumer** [reply]
You can try my "force_url" branch at https://github.com/markus-oberhumer-forks/folke--lazy.nvim .

Indeed "dev = true" does not work for my use case.

**Reply by rameshsanth** [reply]
@amaanq , you're right about cloning a local workspace to develop. I see how it is very easy iff you're working on a single computer.

Here are couple of reasons why private repo support is needed.
1. Multiple servers. I develop on one server, but do have to login into other servers for work. My vim config is in repo. My configs are  synced automatically to other servers by a simple git push/pull. Private repo makes this much easier without having to create multiple workspaces in every server.
2. Private changes becoming permanent. Often PR created for development don't get accepted/merged. Therefore your changes end up in your private repo permanently.

@markus-oberhumer , I applied your patch (using git am) and it works. Unfortunately i found that lazyvim updates overwrite the private changes. so i find myself reapplying the patch after every update. a minor inconvenience.

**Reply by rameshsanth** [reply]
@Jomik I understand your concern how private repo's can deviate from the main repo and cause issues with LazyVims configuration. Onus of keeping the forked repo in sync with main repo is on the developer who created the fork.
You can also make the same case anytime a developer uses a private repo (using "dev" option or forked repo).
Theoretically the developer can make changes in the private repo that breaks the current lazyvim's configuration.
IMHO, anytime a developer starts using private repo, all bets of supporting that plugin's configuration by LazyVim community is off.

**Reply by markus-oberhumer** [reply]
This is what I'm using (diff is against https://github.com/LazyVim/starter )

@rameshsanth As you can see I forceurl lazy.nvim itself, so updates work fine

```diff
diff --git a/lua/config/lazy.lua b/lua/config/lazy.lua
index 891b190..9c49a60 100644
--- a/lua/config/lazy.lua
+++ b/lua/config/lazy.lua
@@ -1,13 +1,34 @@
 local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
 if not vim.loop.fs_stat(lazypath) then
   -- bootstrap lazy.nvim
+  -- vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", "--branch=stable", lazypath })
+  -- using fork with force_url support
   -- stylua: ignore
-  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", "--branch=stable", lazypath })
+  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/markus-oberhumer-forks/folke--lazy.nvim", "--branch=force_url", lazypath })
 end
 vim.opt.rtp:prepend(vim.env.LAZY or lazypath)

 require("lazy").setup({
   spec = {
+    -- using forks with force_url
+    {
+      "folke/lazy.nvim",
+      name = "lazy.nvim",
+      force_url = "https://github.com/markus-oberhumer-forks/folke--lazy.nvim",
+      force_branch = "force_url",
+    },
+    {
+      "nvim-lualine/lualine.nvim",
+      name = "lualine.nvim",
+      force_url = "https://github.com/markus-oberhumer-forks/nvim-lualine--lualine.nvim",
+      force_branch = "mfx",
+    },
+    {
+      "nvim-treesitter/nvim-treesitter",
+      name = "nvim-treesitter",
+      force_url = "https://github.com/markus-oberhumer-forks/nvim-treesitter--nvim-treesitter",
+      force_branch = "mfx",
+    },
     -- add LazyVim and import its plugins
     { "LazyVim/LazyVim", import = "lazyvim.plugins" },
     -- import any extras modules here
```

**Comment by markus-oberhumer-forks** [comment, marked as answer]
Just in case somebody else has a similar problem,  the following quick hack in lazy.nvim seems to work for me:

```diff
diff --git a/lua/lazy/core/plugin.lua b/lua/lazy/core/plugin.lua
index a0b1428..fbe9fa3 100644
--- a/lua/lazy/core/plugin.lua
+++ b/lua/lazy/core/plugin.lua
@@ -291,7 +291,8 @@ end
 function Spec:merge(old, new)
   new._.dep = old._.dep and new._.dep

-  if new.url and old.url and new.url ~= old.url then
+  local force_url = new.force_url
+  if not force_url and new.url and old.url and new.url ~= old.url then
     self:error("Two plugins with the same name and different url:\n" .. vim.inspect({ old = old, new = new }))
   end

@@ -302,6 +303,9 @@ function Spec:merge(old, new)
   new._.super = old
   setmetatable(new, { __index = old })

+  if force_url then
+    new.url = force_url
+  end
   return new
 end
```

**Comment by markus-oberhumer-forks** [comment]
Updated "force_url" fork now available here: https://github.com/markus-oberhumer-forks/folke--lazy.nvim

**Comment by abeldekat** [comment]
Hello @markus-oberhumer-forks,

I read this discussion with interest and I am wondering:
Would a simple `ansible` script solve the use case you describe?

If I understand you correctly, there are two factors at play:
1. Developing plugins, supported by [lazy.nvim](https://github.com/LazyVim/LazyVim/discussions/452#discussioncomment-5352386).
2. Distributing said plugins across machines. This is where `ansible` shines.

**Reply by markus-oberhumer** [reply]
@abeldekat You can use `ansible` to distribute/manage your vim configs, but the problem I'm solving here is unrelated.
